### PR TITLE
Default index mode for chainerx `Take`

### LINKED ISSUE #8276: ChainerX example cannot build
#8197 introduced a mandatory mode argument to `chx::Take()` and `chx::ndarray::Take()`.
This has made the MNIST example unable to build.

https://github.com/chainer/chainer/blob/4fd2210c16b32408d399fa47ae4d29daa9c038a4/chainerx_cc/examples/mnist/train_mnist.cc#L123-L124

I think the argument should be optional.

### PR DESCRIPTION
**Pull request by emcastillo**
Allows the examples to be built

Closes #8276

### CODE REVIEW

File: chainerx_cc/chainerx/array.h [hunk 1]
```diff
@@ -204,7 +204,7 @@ class Array {
     // TODO(niboshi): Support Scalar and StackVector as indices.
     // TODO(niboshi): Support axis=None behavior in NumPy.
     // TODO(niboshi): Support indices dtype other than int64.
-    Array Take(const Array& indices, int8_t axis, IndexBoundsMode mode) const;
+    Array Take(const Array& indices, int8_t axis, const absl::optional<IndexBoundsMode>& mode = absl::nullopt) const;
```
Comment (niboshi) [inline chainerx_cc/chainerx/array.h]: How about this?
```suggestion
    Array Take(const Array& indices, int8_t axis, IndexBoundsMode mode = IndexBoundsMode::kDefault) const;
```

Comment (emcastillo) [reply chainerx_cc/chainerx/array.h]: Yeah, I don't know what I was thinking ...

Comment (niboshi) [reply chainerx_cc/chainerx/array.h]: Maybe because I wrote "optional" in #8276. 😅

Comment (emcastillo) [reply chainerx_cc/chainerx/array.h]: Actually there is a circular reference with routines/indexing.h where the IndexBoundsMode is defined and array.h 😓 preventing to access members of IndexBoundsMode from array.h

Comment (emcastillo) [reply chainerx_cc/chainerx/array.h]: indexing.h uses the Array class, so using array_fwd does not solve the issue 😔
However, I found there is an array_index.h header with all the array indexing related information, so moving the IndexBoundsMode there seems to be a natural choice that solve the issues.
What do you think?

Comment (niboshi) [reply chainerx_cc/chainerx/array.h]: >indexing.h uses the complete Array class

`indexing.h` only includes prototypes. I don't think the complete class is needed.

**Comment by emcastillo** [comment]
PTAL

File: chainerx_cc/chainerx/array_index.h [hunk 2]
```diff
@@ -16,6 +16,14 @@ enum class ArrayIndexTag {
     kEllipsis,
 };
 
+enum class IndexBoundsMode {
+    // Index out-of-bounds handling modes for take
```
Comment (emcastillo) [inline chainerx_cc/chainerx/array_index.h]: TODO: remove this comment or update it with all the routines that support it
`choose` also allows index mode kwarg

**Comment by emcastillo** [comment]
I updated the comment that describes the IndexBoundsMode,
Since this indexing mode is used at more than one routine in numpy, I think that the most appropiate place to declare it is the `array_index` file as I discussed with @niboshi before.

Should we merge this instead of #8288?

**Comment by hvy** [comment]
This basically LGTM, @niboshi may I proceed with merging it?

File: chainerx_cc/chainerx/array_index.h [hunk 3]
```diff
@@ -16,6 +16,15 @@ enum class ArrayIndexTag {
     kEllipsis,
 };
 
+enum class IndexBoundsMode {
+    // Index out-of-bounds handling modes for
```
Comment (niboshi) [inline chainerx_cc/chainerx/array_index.h]: Please place the comment outside the definition.

File: chainerx_cc/chainerx/routines/indexing.h [hunk 4]
```diff
@@ -8,7 +8,6 @@
 
 namespace chainerx {
 namespace internal {
-
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/indexing.h]: Undo this change.

File: chainerx_cc/chainerx/routines/indexing.cc [hunk 5]
```diff
@@ -142,7 +142,6 @@ Array AddAt(const Array& a, const Array& indices, int8_t axis, const Array& b, I
     CHAINERX_ASSERT(0 <= axis && axis < a.ndim());
     CHAINERX_ASSERT(b.ndim() == indices.ndim() + a.ndim() - 1);
     CheckEqual(a.dtype(), b.dtype());
-
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/indexing.cc]: Ditto.
By the way, as `AddAt` is now public, asserts should be exceptions instead (not to be done in this PR though).

**Comment by emcastillo** [comment]
Comments addressed! thanks a lot :)!

**Comment by niboshi** [comment]
Thanks
Jenkins, test this please
